components/sys/drivers/can.c

```diff
@@ -145,26 +147,41 @@ driver_error_t *can_check_error(esp_err_t error) {
  * Helper functions
  */
 
+#define MAX_BUS_ERROR 127
+#define MAX_WAITMS_TX 100
+
+void can_recovery() {
+    can_status_info_t status_info;
+    esp_err_t error = can_get_status_info(&status_info);
+    if (error == ESP_OK && status_info.state==CAN_STATE_RUNNING && status_info.bus_error_count<MAX_BUS_ERROR) {
+        ; //do nothing
+    } else {
+        can_start();
+        can_initiate_recovery();
+    }
+}
+
 static driver_error_t *can_ll_tx(can_message_t *frame) {
-    can_start(); //make sure can is running
-    can_initiate_recovery(); //make sure the bus is on
     driver_error_t *error;
-    if ((error = can_check_error(can_transmit(frame, portMAX_DELAY)))) return error;
+    if ((error = can_check_error(can_transmit(frame, pdMS_TO_TICKS(MAX_WAITMS_TX))))) {
+        can_recovery();
+        return error;
+    }
     return 0;
 }
 
 static driver_error_t *can_ll_rx(can_message_t *frame) {
-    can_start(); //make sure can is running
-    can_initiate_recovery(); //make sure the bus is on
-
```

Comment:
these calls should not be performed on every send so I have moved them to the error-case only